File: src/main/python/Tries/trie.py
from __future__ import annotations


class TrieNode:
    def __init__(self, end_char: bool = False) -> None:
        self.map: dict[str, TrieNode] = {}
        self.end_char = end_char

    def get(self, char: str) -> TrieNode | None:
        return self.map.get(char)

    def get_or_modify(self, char: str) -> TrieNode:
        if char not in self.map:
            self.map[char] = TrieNode()
        return self.map[char]


class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        curr_node = self.root

        for char in word:
            curr_node = curr_node.get_or_modify(char)

        curr_node.end_char = True

    def search(self, word: str) -> bool:
        curr_node = self.root

        for char in word:
            if curr_node is None:
                return False
            curr_node = curr_node.get(char)

        return curr_node is not None and curr_node.end_char

    def prefix_search(self, prefix: str) -> bool:
        curr_node = self.root

        for char in prefix:
            if curr_node is None:
                return False
            curr_node = curr_node.get(char)

        return curr_node is not None

File: src/main/java/Tries/Trie.java
package main.java.Tries;

import java.util.Map;
import java.util.HashMap;

class TrieNode {
    private Map<Character, TrieNode> map;
    boolean isEnd;

    public TrieNode() {
        this.map = new HashMap<>();
        this.isEnd = false;
    }

    public TrieNode getOrModify(char c) {
        return this.map.computeIfAbsent(c, k -> new TrieNode());
    }

    public TrieNode get(char c) {
        return this.map.get(c);
    }
}

public class Trie {
    TrieNode root;

    public Trie() {
        this.root = new TrieNode();
    }

    public void insert(String word) {          
        TrieNode currNode = this.root;
        for (int i = 0; i < word.length(); i++) {
            currNode = currNode.getOrModify(word.charAt(i));
        } currNode.isEnd = true;
    }

    public boolean search(String word) {
        TrieNode currNode = this.root;
        for (int i = 0; i < word.length() && currNode != null; i++) {
            currNode = currNode.get(word.charAt(i));
        }

        if (currNode == null)
            return false;
        return currNode.isEnd;
    }

    public boolean prefixSearch(String prefix) {
        TrieNode currNode = this.root;
        for (int i = 0; i < prefix.length() && currNode != null; i++) {
            currNode = currNode.get(prefix.charAt(i));
        }

        return currNode != null;
    }
}
